src/coreclr/jit/utils.h

```diff
@@ -779,13 +780,128 @@ namespace CheckedOps
 const bool Unsigned = true;
 const bool Signed   = false;
 
-bool IntAddOverflows(int32_t firstAddend, int32_t secondAddend, bool unsignedAdd);
-bool IntSubOverflows(int32_t minuend, int32_t subtrahend, bool unsignedSub);
-bool IntMulOverflows(int32_t firstFactor, int32_t secondFactor, bool unsignedMul);
+// Important note: templated functions below must use dynamic "assert"s instead of "static_assert"s
+// because they can be instantiated on code paths that are not reachable at runtime, but visible
+// to the compiler. One example is VN's EvalOp<T> function, which can be instantiated with "size_t"
+// for some operators, and that's legal, but its callee EvalOpSpecialized<T> uses "assert(!AddOverflows(v1, v2))"
+// for VNF_ADD_OVF/UN, and would like to continue doing so without casts.
 
-bool LongAddOverflows(int64_t firstAddend, int64_t secondAddend, bool unsignedAddition);
-bool LongSubOverflows(int64_t minuend, int64_t subtrahend, bool unsignedSubtraction);
-bool LongMulOverflows(int64_t firstFactor, int64_t secondFactor, bool unsignedMultiplication);
+template <class T>
+bool AddOverflows(T firstAddend, T secondAddend, bool unsignedAdd)
```

Comment:
Have you considered using Clang's built-in functions (`__builtin_sadd_overflow` and friends) as they should be more performant? https://wiki.sei.cmu.edu/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow

Reply:
I guess we still need to support building with VS, are you suggesting different code paths for VS and clang?

Reply:
We might use `#ifdef __GNUC__` if the code is properly structured.

Reply:
Yes. I decided against it for simplicity reasons - this code is quite cold as checked operations are rare.